File: wear/src/main/java/com/garethevans/church/opensongtablet/MetronomeListenerService.java
package com.garethevans.church.opensongtablet;

import android.content.Intent;
import android.graphics.Color;
import android.os.Build;
import android.os.Looper;
import android.os.VibrationEffect;
import android.os.Vibrator;
import android.util.Log;

import androidx.annotation.NonNull;

import com.google.android.gms.wearable.DataEvent;
import com.google.android.gms.wearable.DataEventBuffer;
import com.google.android.gms.wearable.DataMap;
import com.google.android.gms.wearable.DataMapItem;
import com.google.android.gms.wearable.MessageEvent;
import com.google.android.gms.wearable.WearableListenerService;

/**
 * Listens for metronome beat messages from the phone and vibrates the watch in time.
 * No deprecated APIs; compatible with Play Services 18+ and Wear OS 4+.
 */
public class MetronomeListenerService extends WearableListenerService {

    private static final String TAG = "MetronomeService";
    public static final String BEAT_PATH_TICK = "/metronome/beat/tick";
    public static final String BEAT_PATH_TOCK = "/metronome/beat/tock";
    private static final String METRONOME_STATE_PATH = "/metronome/state";
    public static final String ACTION_METRONOME_SYNC = "com.garethevans.church.opensongtablet.METRONOME_SYNC";
    @SuppressWarnings("FieldCanBeLocal")
    private final int tickLength = 50;
    @SuppressWarnings("FieldCanBeLocal")
    private final int tockLength = 30;
    @SuppressWarnings("FieldCanBeLocal")
    private final int tickAmplitude = 255;
    @SuppressWarnings("FieldCanBeLocal")
    private final int tockAmplitude = 120;

    private Vibrator vibrator;

    @Override
    public void onCreate() {
        super.onCreate();
        vibrator = (Vibrator) getSystemService(VIBRATOR_SERVICE);
        Log.d(TAG, "Listener service created (emulator-friendly)");
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        // Optional: trigger a test beat from intent extras
        if (intent != null && intent.hasExtra("test_beat")) {
            String path = intent.getStringExtra("test_beat");
            simulateBeat(path);
        }
        return START_STICKY;
    }

    @Override
    public void onDataChanged(DataEventBuffer dataEvents) {
        for (DataEvent event : dataEvents) {
            if (event.getType() == DataEvent.TYPE_CHANGED &&
                    METRONOME_STATE_PATH.equals(event.getDataItem().getUri().getPath())) {

                DataMap dataMap = DataMapItem.fromDataItem(event.getDataItem()).getDataMap();
                boolean isPlaying = dataMap.getBoolean("isPlaying");
                int bpm = dataMap.getInt("bpm");
                int beatsPerBar = dataMap.getInt("beatsPerBar", 4);
                int beatDenominator = dataMap.getInt("beatDenominator", 4); // Pulls 4, 8, etc.

                Log.d(TAG, "Sync Event from Phone -> Playing: " + isPlaying + ", BPM: " + bpm + ", Signature: " + beatsPerBar + "/4");

                Intent intent = new Intent(ACTION_METRONOME_SYNC);
                intent.putExtra("isPlaying", isPlaying);
                intent.putExtra("bpm", bpm);
                intent.putExtra("beatsPerBar", beatsPerBar);
                intent.putExtra("beatDenominator", beatDenominator);


                //Intent intent = new Intent("com.opensong.METRONOME_UPDATE");
                //intent.putExtra("isPlaying", isPlaying);
                //intent.putExtra("bpm", bpm);
                //intent.putExtra("beatsPerBar", beatsPerBar);
                //intent.putExtra("beatDenominator", beatDenominator);
                sendBroadcast(intent);
                Log.d("MetronomeService", "📣 Broadcast dispatched on channel: " + ACTION_METRONOME_SYNC);
            }
        }
    }

    /**
     * Call this to simulate a beat for emulator testing
     */
    public void simulateBeat(String path) {
        Log.d(TAG, "Simulating beat: " + path);
        if (BEAT_PATH_TICK.equals(path)) {
            triggerVibration(true);  // main beat
        } else if (BEAT_PATH_TOCK.equals(path)) {
            triggerVibration(false); // offbeat
        } else {
            Log.d(TAG, "Unknown beat path: " + path);
        }
    }

    private void triggerVibration(boolean mainBeat) {
        if (vibrator == null) return;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            VibrationEffect effect = mainBeat
                    ? VibrationEffect.createOneShot(tickLength, tickAmplitude)
                    : VibrationEffect.createOneShot(tockLength, tockAmplitude); // weaker offbeat
            vibrator.vibrate(effect);
        } else {
            vibrator.vibrate(mainBeat ? 50 : 30);
        }
        Log.d(TAG, "Vibration triggered: " + (mainBeat ? "TICK" : "TOCK"));
    }


    @Override
    public void onMessageReceived(@NonNull MessageEvent event) {
        Log.d(TAG, "messageReceived");
        if (BEAT_PATH_TICK.equals(event.getPath())) {
            triggerVibration(50, 255);
        } else if (BEAT_PATH_TOCK.equals(event.getPath())) {
            triggerVibration(30, 120);
        } else {
            Log.d(TAG, "Unknown message path: " + event.getPath());
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.d(TAG, "Listener service destroyed");
    }

    private void triggerVibration(int length, int amplitude) {
        Log.d(TAG, "Vibrate triggered");
        Vibrator vibrator = (Vibrator) getSystemService(VIBRATOR_SERVICE);
        if (vibrator != null && vibrator.hasVibrator()) {
            // Short, crisp vibration for each beat
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                vibrator.vibrate(VibrationEffect.createOneShot(length, amplitude));
            } else {
                vibrator.vibrate(length);  // fallback for older devices
            }
            Log.d(TAG, "Vibrate triggered");

        } else {
            Log.w(TAG, "No vibrator available on this device");
        }
    }



}

File: wear/src/main/kotlin/com/garethevans/church/opensongtablet/MetronomeListenerService.kt
package com.garethevans.church.opensongtablet

import android.content.Intent
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import com.google.android.gms.wearable.DataEvent
import com.google.android.gms.wearable.DataEventBuffer
import com.google.android.gms.wearable.DataMapItem
import com.google.android.gms.wearable.MessageEvent
import com.google.android.gms.wearable.WearableListenerService

/**
 * Listens for metronome beat messages from the phone and vibrates the watch in time.
 * No deprecated APIs; compatible with Play Services 18+ and Wear OS 4+.
 */
class MetronomeListenerService : WearableListenerService() {
    private val tickLength = 50
    private val tockLength = 30
    private val tickAmplitude = 255
    private val tockAmplitude = 120

    private var vibrator: Vibrator? = null

    @Suppress("DEPRECATION")
    override fun onCreate() {
        super.onCreate()
        vibrator = getSystemService(VIBRATOR_SERVICE) as Vibrator?
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        // Optional: trigger a test beat from intent extras
        if (intent != null && intent.hasExtra("test_beat")) {
            val path = intent.getStringExtra("test_beat")
            simulateBeat(path)
        }
        return START_STICKY
    }

    override fun onDataChanged(dataEvents: DataEventBuffer) {
        for (event in dataEvents) {
            if (event.type == DataEvent.TYPE_CHANGED &&
                METRONOME_STATE_PATH == event.dataItem.uri.path
            ) {
                val dataMap = DataMapItem.fromDataItem(event.dataItem).dataMap
                val isPlaying = dataMap.getBoolean("isPlaying")
                val bpm = dataMap.getInt("bpm")
                val beatsPerBar = dataMap.getInt("beatsPerBar", 4)
                val beatDenominator = dataMap.getInt("beatDenominator", 4) // Pulls 4, 8, etc.

                val intent = Intent(ACTION_METRONOME_SYNC)
                intent.setPackage(this.packageName) // Explicitly target your own app
                intent.putExtra("isPlaying", isPlaying)
                intent.putExtra("bpm", bpm)
                intent.putExtra("beatsPerBar", beatsPerBar)
                intent.putExtra("beatDenominator", beatDenominator)

sendBroadcast(intent)

            }
        }
    }

    /**
     * Call this to simulate a beat for emulator testing
     */
    fun simulateBeat(path: String?) {
        if (BEAT_PATH_TICK == path) {
            triggerVibration(true) // main beat
        } else if (BEAT_PATH_TOCK == path) {
            triggerVibration(false) // offbeat
        }
    }

    @Suppress("DEPRECATION")
    private fun triggerVibration(mainBeat: Boolean) {
        if (vibrator == null) return

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val effect = if (mainBeat)
                VibrationEffect.createOneShot(tickLength.toLong(), tickAmplitude)
            else
                VibrationEffect.createOneShot(tockLength.toLong(), tockAmplitude) // weaker offbeat
            vibrator!!.vibrate(effect)
        } else {
            vibrator!!.vibrate((if (mainBeat) 50 else 30).toLong())
        }
    }

    override fun onMessageReceived(event: MessageEvent) {
        if (BEAT_PATH_TICK == event.path) {
            triggerVibration(50, 255)
        } else if (BEAT_PATH_TOCK == event.path) {
            triggerVibration(30, 120)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    @Suppress("DEPRECATION")
    private fun triggerVibration(length: Int, amplitude: Int) {
        val vibrator = getSystemService(VIBRATOR_SERVICE) as Vibrator?
        if (vibrator != null && vibrator.hasVibrator()) {
            // Short, crisp vibration for each beat
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                vibrator.vibrate(VibrationEffect.createOneShot(length.toLong(), amplitude))
            } else {
                vibrator.vibrate(length.toLong()) // fallback for older devices
            }
        }
    }


    @Suppress("UNUSED")
    companion object {
        private const val TAG = "MetronomeService"
        const val BEAT_PATH_TICK: String = "/metronome/beat/tick"
        const val BEAT_PATH_TOCK: String = "/metronome/beat/tock"
        private const val METRONOME_STATE_PATH = "/metronome/state"
        const val ACTION_METRONOME_SYNC: String =
            "com.garethevans.church.opensongtablet.METRONOME_SYNC"
    }
}
